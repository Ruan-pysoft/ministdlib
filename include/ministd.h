#ifndef _RM_STD_H
#define _RM_STD_H

#include <_ministd_prelude.h>

/* This is the main header for providing basic startup and exit functionality,
 * as well as some basic interaction with the program's environment.
 */

/* These functions are used to access
 * the program's command-line arguments (argc/argv)
 * and the program's environment variables (envc/envp)
 */
int argc(void);
char ref ref argv(void);
int envc(void);
char ref ref envp(void);

/* Functions to terminate the program */
void atexit(void (ref func)(void));
void exit(int exitcode);
void thread_exit(int exitcode);
void panic(isz exitcode);
#define PANIC(code) do { panic(code); __builtin_unreachable(); } while (0)

/* The program's entry point, called by the kernel
 * _start in turn call's the user's main.
 * See https://en.wikipedia.org/wiki/Entry_point#Contemporary */
void _start(void);

#endif /* _RM_STD_H */
