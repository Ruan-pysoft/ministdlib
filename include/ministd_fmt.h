#ifndef _RM_STD_FMT_H
#define _RM_STD_FMT_H

#include "ministd_io.h"
#include "ministd_error.h"
#include "ministd_types.h"

#define FPRINT_RAW_DECL(type, suffix) \
	isz fprint##suffix(type suffix, FILE ref file, err_t ref err_out); \
	isz fprint##suffix##b(type suffix, FILE ref file, err_t ref err_out); \
	isz fprint##suffix##x(type suffix, FILE ref file, err_t ref err_out); \
	isz print##suffix(type suffix, err_t ref err_out); \
	isz print##suffix##b(type suffix, err_t ref err_out); \
	isz print##suffix##x(type suffix, err_t ref err_out)

#define FPRINT_DECL(type, suffix) \
	FPRINT_RAW_DECL(type, suffix); \
	FPRINT_RAW_DECL(unsigned type, u##suffix)

isz fprintc(char c, FILE ref file, err_t ref err_out);
isz fprints(const char ref s, FILE ref file, err_t ref err_out);
isz printc(char c, err_t ref err_out);
isz prints(const char ref s, err_t ref err_out);

FPRINT_DECL(long, l);

FPRINT_RAW_DECL(isz, z);
FPRINT_RAW_DECL(usz, uz);

FPRINT_DECL(int, i);

FPRINT_DECL(short, h);

isz fprintp(const ptr p, FILE ref file, err_t ref err_out);
isz printp(const ptr p, err_t ref err_out);

#undef FPRINT_DECL
#undef FPRINT_RAW_DECL

char fscanc(FILE ref file, err_t ref err_out);
char own fscans(FILE ref file, err_t ref err_out);
long fscanl(FILE ref file, err_t ref err_out);
unsigned long fscanul(FILE ref file, err_t ref err_out);
isz fscanz(FILE ref file, err_t ref err_out);
usz fscanuz(FILE ref file, err_t ref err_out);
int fscani(FILE ref file, err_t ref err_out);
unsigned int fscanu(FILE ref file, err_t ref err_out);
short fscanh(FILE ref file, err_t ref err_out);
unsigned short fscanuh(FILE ref file, err_t ref err_out);

#define scanc(err_out)	fscanc(stdin, err_out)
#define scans(err_out)	fscans(stdin, err_out)
#define scanl(err_out)	fscanl(stdin, err_out)
#define scanul(err_out)	fscanul(stdin, err_out)
#define scanz(err_out)	fscanz(stdin, err_out)
#define scanuz(err_out)	fscanuz(stdin, err_out)
#define scani(err_out)	fscani(stdin, err_out)
#define scanu(err_out)	fscanu(stdin, err_out)
#define scanh(err_out)	fscanh(stdin, err_out)
#define scanuh(err_out)	fscanuh(stdin, err_out)

#endif /* _RM_STD_FMT_H */
