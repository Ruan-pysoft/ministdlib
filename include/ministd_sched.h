#ifndef _RM_STD_SCHED_H
#define _RM_STD_SCHED_H

#include <_ministd_prelude.h>

#include <linux/sched.h>

#define _syscall_clone3(res, cl_args) \
		_syscall2(__NR_clone3, res, cl_args, sizeof(*cl_args));


usz clone(int (ref fn)(ptr), ptr arg, struct clone_args ref cl_args,
	  err_t ref err_out);

#endif /* _RM_STD_SCHED_H */
