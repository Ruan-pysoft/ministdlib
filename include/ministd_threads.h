#ifndef _RM_STD_THREADS_H
#define _RM_STD_THREADS_H

#include <_ministd_prelude.h>

typedef struct AtomicI mutex_t;

mutex_t own mutex_new(err_t ref err_out);
bool mutex_try_lock(mutex_t ref this);
void mutex_lock(mutex_t ref this);
void mutex_unlock(mutex_t ref this);

#endif /* _RM_STD_THREADS_H */
