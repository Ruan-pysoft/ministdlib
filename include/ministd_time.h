#ifndef _RM_STD_TIME_H
#define _RM_STD_TIME_H

#include <_ministd_prelude.h>

struct timespec {
	long tv_sec;
	long tv_nsec;
};

long nanosleep(const struct timespec ref sleep_time,
	       struct timespec ref time_left, err_t ref err_out);
void millisleep(unsigned int millis, err_t ref err_out);
void sleep(float secs, err_t ref err_out);

#endif /* _RM_STD_TIME_H */
