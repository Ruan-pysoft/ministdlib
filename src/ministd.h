#ifndef _RM_STD_H
#define _RM_STD_H

#include "ministd_io.h"
#include "ministd_types.h"

int argc(void);
char ref ref argv(void);
int envc(void);
char ref ref envp(void);

void atexit(void (ref func)(void));
void exit(int exitcode);
void _start(void);

#endif /* _RM_STD_H */
