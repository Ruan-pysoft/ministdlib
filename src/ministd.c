#include "ministd.h"

#include "ministd_syscall.h"

#include <asm/unistd.h>

void
exit(int exitcode)
{
	isz lcode, dummy;
	lcode = exitcode;

	_syscall1(__NR_exit, dummy, lcode);
	__builtin_unreachable();
}
int main(int argc, char ref ref argv);
extern void ministd_io_init(void);
void
setup(void)
{
	ministd_io_init();
}
void
_start(void)
{
	int argc;
	char ref ref argv;

	__asm__ volatile(
		"mov %%rsp, %%rax;" /* sp -> ax */
		"mov 24(%%rax), %[argc];" /* retrieve argc, top item of stack */
		"mov %%rax, %[argv];" /* retrieve argv, second-to-top of stack */
		"add $32, %[argv];"
		: [argc] "=r" (argc), [argv] "=r" (argv) /* outputs */
		: /* inputs */
		: "rax" /* clobber */
	);

	setup();

	exit(main(argc, argv));
	__builtin_unreachable();
}

#include "ministd_io.h"

void
__stack_chk_fail(void)
{
	const char msg[] = "Stack smashing detected!";

	fd_write(2 /* stderr */, (ptr)msg, sizeof(msg)/sizeof(*msg), NULL);
	exit(127);
}
