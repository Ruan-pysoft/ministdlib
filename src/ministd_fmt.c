#include <ministd_fmt.h>

#include <ministd_error.h>
#include <ministd_io.h>
#include <ministd_memory.h>

isz
fprintc(char c, FILE ref file, err_t ref err_out)
{
	err_t err = ERR_OK;
	fputc(c, file, &err);
	TRY_WITH(err, 0);
	return 1;
}
isz
fprints(const char ref s, FILE ref file, err_t ref err_out)
{
	return fputs(s, file, err_out);
}
isz
printc(char c, err_t ref err_out)
{ return fprintc(c, stdout, err_out); }
isz
prints(const char ref s, err_t ref err_out)
{ return fprints(s, stdout, err_out); }

isz
fprintl(long l, FILE ref file, err_t ref err_out)
{
	/* max int is 2^63-1 ~= 10^18 * 8, so no more than 20 digits */
	err_t err = ERR_OK;
	char buf[21];
	isz len = 0;

	if (l == 0) return fprintc('0', file, err_out);
	if ((unsigned long)l == (~(unsigned long)l)+1) {
		/* l == -2^63 */
		fprintc('-', file, &err);
		TRY_WITH(err, 0);
		return fprintul((unsigned long)l, file, err_out);
	}
	if (l < 0) {
		fprintc('-', file, &err);
		TRY_WITH(err, 0);
		return fprintl(-l, file, err_out);
	}

	buf[20] = 0;
	len = 1;
	while (len <= 20 && l > 0) {
		buf[20-len] = '0' + l%10;
		l /= 10;
		++len;
	}
	--len;

	return fputs(buf+20-len, file, err_out);
}
isz
fprintul(unsigned long ul, FILE ref file, err_t ref err_out)
{
	/* max int is 2^64-1 ~= 10^18 * 16, so no more than 21 digits */
	char buf[22];
	isz len = 0;

	if (ul == 0) return fprintc('0', file, err_out);

	buf[21] = 0;
	len = 1;
	while (len <= 21 && ul > 0) {
		buf[21-len] = '0' + ul%10;
		ul /= 10;
		++len;
	}
	--len;

	return fputs(buf+21-len, file, err_out);
}
isz
fprintlb(long l, FILE ref file, err_t ref err_out)
{
	/* max int is 2^63-1 so at most 63 digits */
	err_t err = ERR_OK;
	char buf[64];
	isz len = 0;

	if (l == 0) return fprintc('0', file, err_out);
	if ((unsigned long)l == (~(unsigned long)l)+1) {
		/* l == -2^63 */
		fprintc('-', file, &err);
		TRY_WITH(err, 0);
		return fprintulb((unsigned long)l, file, err_out);
	}
	if (l < 0) {
		fprintc('-', file, &err);
		TRY_WITH(err, 0);
		return fprintlb(-l, file, err_out);
	}

	buf[63] = 0;
	len = 1;
	while (len <= 63 && l > 0) {
		buf[63-len] = '0' + (l&1);
		l >>= 1;
		++len;
	}
	--len;

	return fputs(buf+63-len, file, err_out);
}
isz
fprintulb(unsigned long ul, FILE ref file, err_t ref err_out)
{
	/* max int is 2^64-1  so at most 64 digits */
	char buf[65];
	isz len = 0;

	if (ul == 0) return fprintc('0', file, err_out);

	buf[64] = 0;
	len = 1;
	while (len <= 64 && ul > 0) {
		buf[64-len] = '0' + (ul&1);
		ul >>= 1;
		++len;
	}
	--len;

	return fputs(buf+64-len, file, err_out);
}
isz
fprintlx(long l, FILE ref file, err_t ref err_out)
{
	/* max int is 2^63-1 so at most 16 digits */
	err_t err = ERR_OK;
	char buf[17];
	isz len = 0;

	if (l == 0) return fprintc('0', file, err_out);
	if ((unsigned long)l == (~(unsigned long)l)+1) {
		/* l == -2^63 */
		fprintc('-', file, &err);
		TRY_WITH(err, 0);
		return fprintulx((unsigned long)l, file, err_out);
	}
	if (l < 0) {
		fprintc('-', file, &err);
		TRY_WITH(err, 0);
		return fprintlx(-l, file, err_out);
	}

	buf[16] = 0;
	len = 1;
	while (len <= 16 && l > 0) {
		int digit = l&0xF;
		if (digit < 10) buf[16-len] = '0' + digit;
		else buf[16-len] = 'a' - 10 + digit;
		l >>= 4;
		++len;
	}
	--len;

	return fputs(buf+16-len, file, err_out);
}
isz
fprintulx(unsigned long ul, FILE ref file, err_t ref err_out)
{
	/* max int is 2^64-1  so at most 16 digits */
	char buf[17];
	isz len = 0;

	if (ul == 0) return fprintc('0', file, err_out);

	buf[16] = 0;
	len = 1;
	while (len <= 16 && ul > 0) {
		int digit = ul&0xF;
		if (digit < 10) buf[16-len] = '0' + digit;
		else buf[16-len] = 'a' - 10 + digit;
		ul >>= 4;
		++len;
	}
	--len;

	return fputs(buf+16-len, file, err_out);
}
isz
printl(long l, err_t ref err_out)
{ return fprintl(l, stdout, err_out); }
isz
printul(unsigned long ul, err_t ref err_out)
{ return fprintul(ul, stdout, err_out); }
isz
printlb(long l, err_t ref err_out)
{ return fprintlb(l, stdout, err_out); }
isz
printulb(unsigned long ul, err_t ref err_out)
{ return fprintulb(ul, stdout, err_out); }
isz
printlx(long l, err_t ref err_out)
{ return fprintlx(l, stdout, err_out); }
isz
printulx(unsigned long ul, err_t ref err_out)
{ return fprintulx(ul, stdout, err_out); }

#define FPRINT_RAW_S_IMPL(type, suffix) \
	isz fprint##suffix(type suffix, FILE ref file, err_t ref err_out) \
	{ return fprintl((long)suffix, file, err_out); } \
	isz fprint##suffix##b(type suffix, FILE ref file, err_t ref err_out) \
	{ return fprintlb((long)suffix, file, err_out); } \
	isz fprint##suffix##x(type suffix, FILE ref file, err_t ref err_out) \
	{ return fprintlx((long)suffix, file, err_out); } \
	isz print##suffix(type suffix, err_t ref err_out) \
	{ return printl((long)suffix, err_out); } \
	isz print##suffix##b(type suffix, err_t ref err_out) \
	{ return printl((long)suffix, err_out); } \
	isz print##suffix##x(type suffix, err_t ref err_out) \
	{ return printl((long)suffix, err_out); }
#define FPRINT_RAW_U_IMPL(type, suffix) \
	isz fprint##suffix(type suffix, FILE ref file, err_t ref err_out) \
	{ return fprintul((unsigned long)suffix, file, err_out); } \
	isz fprint##suffix##b(type suffix, FILE ref file, err_t ref err_out) \
	{ return fprintulb((unsigned long)suffix, file, err_out); } \
	isz fprint##suffix##x(type suffix, FILE ref file, err_t ref err_out) \
	{ return fprintulx((unsigned long)suffix, file, err_out); } \
	isz print##suffix(type suffix, err_t ref err_out) \
	{ return printul((unsigned long)suffix, err_out); } \
	isz print##suffix##b(type suffix, err_t ref err_out) \
	{ return printul((unsigned long)suffix, err_out); } \
	isz print##suffix##x(type suffix, err_t ref err_out) \
	{ return printul((unsigned long)suffix, err_out); }
#define FPRINT_IMPL(type, suffix) \
	FPRINT_RAW_S_IMPL(type, suffix) \
	FPRINT_RAW_U_IMPL(unsigned type, u##suffix)

FPRINT_RAW_S_IMPL(isz, z)
FPRINT_RAW_U_IMPL(usz, uz)

FPRINT_IMPL(int, i)

FPRINT_IMPL(short, h)

isz
fprintp(const ptr p, FILE ref file, err_t ref err_out)
{ return fprintuzx((usz)p, file, err_out); }
isz
printp(const ptr p, err_t ref err_out)
{ return printuzx((usz)p, err_out); }

char
fscanc(FILE ref file, err_t ref err_out)
{
	return fgetc(file, err_out);
}
char own
fscans(FILE ref file, err_t ref err_out)
{
	err_t err = ERR_OK;
	char own result;
	usz result_cap;

	result_cap = 128;
	result = alloc(result_cap, &err);
	TRY_WITH(err, NULL);

	if (fgets(result, result_cap, file, &err)) {
		return result;
	}

	for (;;) {
		if (err != ERR_OK) {
			free(result);
			ERR_WITH(err, NULL);
		}

		result = realloc(result, 2*result_cap, &err);
		TRY_WITH(err, NULL);

		if (fgets(result+result_cap, result_cap, file, &err)) {
			return result;
		}
		result_cap *= 2;
	}
}
#define INTSCAN(type) do {                                           \
		err_t err = ERR_OK;                                  \
		char c;                                              \
		type res = 0;                                        \
                                                                     \
		while ((c = peekc(file, &err) != -1) && c <= ' ') {  \
			TRY_WITH(err, 0);                            \
			fgetc(file, &err);                           \
			TRY_WITH(err, 0);                            \
		}                                                    \
		TRY_WITH(err, 0);                                    \
		if (c == -1) ERR_WITH(ERR_EOF, 0);                   \
                                                                     \
		if (c < '0' || c > '9') ERR_WITH(ERR_PARSE, 0);      \
		while ((c = fgetc(file, &err)) >= '0' && c <= '9') { \
			const type oldres = res;                     \
			TRY_WITH(err, 0);                            \
                                                                     \
			res *= 10;                                   \
			res += c - '0';                              \
                                                                     \
			if (res < oldres) {                          \
				ERR_WITH(ERR_OVERFLOW, 0);           \
			}                                            \
		}                                                    \
		TRY_WITH(err, 0);                                    \
		if (c != -1) {                                       \
			ungetc(file, c, &err);                       \
			TRY_WITH(err, 0);                            \
		}                                                    \
                                                                     \
		return res;                                          \
	} while (0)
long
fscanl(FILE ref file, err_t ref err_out)
{
	INTSCAN(signed long);
}
unsigned long
fscanul(FILE ref file, err_t ref err_out)
{
	INTSCAN(unsigned long);
}
isz
fscanz(FILE ref file, err_t ref err_out)
{
	INTSCAN(isz);
}
usz
fscanuz(FILE ref file, err_t ref err_out)
{
	INTSCAN(usz);
}
int
fscani(FILE ref file, err_t ref err_out)
{
	INTSCAN(signed int);
}
unsigned int
fscanu(FILE ref file, err_t ref err_out)
{
	INTSCAN(unsigned int);
}
short
fscanh(FILE ref file, err_t ref err_out)
{
	INTSCAN(signed short);
}
unsigned short
fscanuh(FILE ref file, err_t ref err_out)
{
	INTSCAN(unsigned short);
}
