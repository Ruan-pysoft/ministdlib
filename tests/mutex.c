#include <linux/sched.h>
#include <ministd.h>
#include <ministd_threads.h>
#include <ministd_fmt.h>
#include <ministd_memory.h>
#include <ministd_time.h>
#include <ministd_sched.h>

#define N_THREADS 10

bool has_executed[N_THREADS] = {0};
volatile usz num_executed = 0;

struct thread_fn_arg {
	mutex_t ref lock;
	int thread_id;
};

int thread_fn(void ref void_arg) {
	struct thread_fn_arg ref arg = void_arg;

	mutex_lock(arg->lock);
	fprints("Acquired lock in thread ", stderr, NULL);
	fprinti(arg->thread_id, stderr, NULL);
	fprints("!\n", stderr, NULL);

	++num_executed;
	has_executed[arg->thread_id] = true;

	sleep(0.1, NULL);
	mutex_unlock(arg->lock);

	return 0;
}

const char ref MSG_LOWER = "hello, world\n";
const char lowercase_bit = 'a'^'A';

void
main(err_t ref err_out)
{
	mutex_t own lock = mutex_new(NULL);
	struct clone_args cl_args = { 0 };
	usz i;

	(void) err_out;

	cl_args.flags = CLONE_VM | CLONE_THREAD | CLONE_FILES | CLONE_FS | CLONE_IO | CLONE_SIGHAND | CLONE_SYSVSEM;

	for (i = 0; i < N_THREADS; ++i) {
		own_ptr stack = alloc(256, NULL);
		struct thread_fn_arg own arg = alloc(sizeof(struct thread_fn_arg), NULL);
		cl_args.stack = (usz)stack;
		cl_args.stack_size = 256;
		arg->lock = lock;
		arg->thread_id = i;

		prints("Launching thread #", NULL);
		printuz(i, NULL);
		prints("...\n", NULL);

		clone(thread_fn, arg, &cl_args, NULL);
	}

	while (num_executed < N_THREADS) {}
	
	prints("All threads have executed! Checking...\n", NULL);

	for (i = 0; i < N_THREADS; ++i) {
		prints("Thread ", NULL);
		printuz(i, NULL);
		prints(" has executed? ", NULL);
		printi(has_executed[i], NULL);
		printc('\n', NULL);
	}
}
