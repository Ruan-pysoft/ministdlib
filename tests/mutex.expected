Launching thread #0...
Launching thread #1...
Launching thread #2...
Launching thread #3...
Launching thread #4...
Launching thread #5...
Launching thread #6...
Launching thread #7...
Launching thread #8...
Launching thread #9...
All threads have executed! Checking...
Thread 0 has executed? 1
Thread 1 has executed? 1
Thread 2 has executed? 1
Thread 3 has executed? 1
Thread 4 has executed? 1
Thread 5 has executed? 1
Thread 6 has executed? 1
Thread 7 has executed? 1
Thread 8 has executed? 1
Thread 9 has executed? 1
